### PR TITLE
Replace Transpose with Replace if they are equivalent

### PR DESCRIPTION
**Pull request by guyang3532**
### Description
Transpose is equivalent to a Reshape if:
 empty dimensions can change place, not empty dimensions must be in
 the same order in the permuted tenosr.
 Example: Shape=(1,1,1024,4096) -> perm=(2,0,3,1).
This pr adds a graph transformer which replaces Transpose with Reshape if they are equivalent.
Because Transpose need memory copy while Reshape needn't, this replacement can save overhead for memory copy.

### CODE REVIEW

File: orttraining/orttraining/core/optimizer/transpose_replacement..cc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#include "core/common/logging/logging.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status TransposeReplacement::Apply(Graph& graph, Node& transpose_node, RewriteRuleEffect& rule_effect, const logging::Logger& logger) const {
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/transpose_replacement..cc:14]: nit: wrap the line.

File: orttraining/orttraining/core/optimizer/transpose_replacement..cc [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  auto& transpose_inputs = transpose_node.MutableInputDefs();
+  auto& transpose_outputs = transpose_node.MutableOutputDefs();
+  NodeArg* input = transpose_inputs[0];
+  auto input_shape = input->Shape();
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/transpose_replacement..cc:18]: early return if input_shape is nullptr

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc [hunk 3]
```diff
@@ -551,6 +552,47 @@ TEST_F(GraphTransformationTests, ConcatReplacement) {
...
+    ASSERT_STATUS_OK(graph_transformation_mgr.Register(std::move(rule_transformer_L1), TransformerLevel::Level1));
+
+    ASSERT_STATUS_OK(graph_transformation_mgr.ApplyTransformers(graph, TransformerLevel::Level1, *logger_));
+
+    std::map<std::string, int> op_to_count = CountOpsInGraph(graph);
+
+    ASSERT_EQ(op_to_count["Transpose"], 1);
+    ASSERT_EQ(op_to_count["Reshape"], 0);
+  }
+
```
Comment (pengwa) [inline orttraining/orttraining/test/optimizer/graph_transform_test.cc:593]: nit: fix the cpplint warning.

File: onnxruntime/test/testdata/transform/transpose_graph_gen.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import numpy as np
```
Comment (pengwa) [inline onnxruntime/test/testdata/transform/transpose_graph_gen.py:1]:
```
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------
```

Comment (pengwa) [inline onnxruntime/test/testdata/transform/transpose_graph_gen.py:1]: Import of 'np' is not used.

**Comment by pengwa** [comment]
You can run `lintrunner -a`  locally to found all the pylint issues.

File: orttraining/orttraining/core/optimizer/transpose_replacement.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/compute_optimizer/shared_utils.h"
+
+namespace onnxruntime {
+
+/**
+@Class TransposeReplacement
+
+Transpose is equivalent to a Reshape if:
+ empty dimensions can change place, not empty dimensions must be in
```
[the remark this answers is not in the document]
Comment (guyang3532) [reply orttraining/orttraining/core/optimizer/transpose_replacement.h]: added, thx

File: orttraining/orttraining/core/optimizer/transpose_replacement..cc (hunk 1 above)
Comment (guyang3532) [reply orttraining/orttraining/core/optimizer/transpose_replacement..cc]: fixed, thx

File: orttraining/orttraining/core/optimizer/transpose_replacement..cc (hunk 2 above)
Comment (guyang3532) [reply orttraining/orttraining/core/optimizer/transpose_replacement..cc:21]: fixed, thx

File: onnxruntime/test/testdata/transform/transpose_graph_gen.py (hunk 4 above)
Comment (guyang3532) [reply onnxruntime/test/testdata/transform/transpose_graph_gen.py]: fixed, thx

File: orttraining/orttraining/test/optimizer/graph_transform_test.cc (hunk 3 above)
Comment (guyang3532) [reply orttraining/orttraining/test/optimizer/graph_transform_test.cc]: fixed, thx

File: onnxruntime/test/testdata/transform/transpose_graph_gen.py (hunk 4 above)
Comment (guyang3532) [reply onnxruntime/test/testdata/transform/transpose_graph_gen.py]: removed, thx

**Review by pengwa — APPROVED** [review]

**Pull request merged**